File: mobile/api/savedFoods.ts
/**
 * Typed client for the FTY-052 saved-food save and typeahead search API.
 *
 * Two operations consumed by the mobile saved-food UI (FTY-053):
 *
 * - `saveFood` — `POST /api/users/{user_id}/saved-foods`: persists a corrected
 *   food from the editing surface (FTY-050), recording the typed phrase as an
 *   alias. The caller supplies the canonical name, the originating phrase, and
 *   the corrected nutrition snapshot.
 * - `searchSavedFoods` — `GET /api/users/{user_id}/saved-foods?q=…`: prefix-
 *   matched typeahead against the user's saved foods and their aliases,
 *   returning the matching foods with stored nutrition so the client can apply
 *   them directly.
 *
 * Both operations are scoped to the authenticated caller: cross-user access
 * fails closed as 404. The typed phrase and query text are never logged here;
 * errors carry only the HTTP status and the attempted action.
 */

/** Per-serving nutrition crossing the save boundary (mirrors FTY-052 backend). */
export interface NutritionSnapshot {
  readonly calories: number;
  readonly protein_g: number | null;
  readonly carbs_g: number | null;
  readonly fat_g: number | null;
  readonly serving_size: number;
  readonly serving_unit: string;
}

/** Request body for the deliberate save of a corrected food. */
export interface SaveFoodRequest {
  /** Canonical name to save under (also used as the display name). */
  readonly name: string;
  /** The typed phrase that originated this save, persisted as an alias. */
  readonly phrase: string;
  /** The corrected nutrition snapshot to store. */
  readonly nutrition: NutritionSnapshot;
}

/** A user-owned saved food returned on save and typeahead search. */
export interface SavedFoodDTO {
  readonly id: string;
  readonly user_id: string;
  readonly name: string;
  readonly calories: number;
  readonly protein_g: number | null;
  readonly carbs_g: number | null;
  readonly fat_g: number | null;
  readonly serving_size: number;
  readonly serving_unit: string;
  /** Provenance: always `'saved_from_correction'` in v1. */
  readonly source: string;
  readonly created_at: string;
  readonly updated_at: string;
}

/** Bounded typeahead response: matched saved foods and the applied limit. */
export interface SavedFoodSearchResponse {
  readonly items: readonly SavedFoodDTO[];
  readonly limit: number;
}

/** Authenticated session needed to address the owner's saved foods. */
export interface SavedFoodSession {
  readonly baseUrl: string;
  readonly token: string;
  readonly userId: string;
}

/** Raised when the saved-food API returns a non-2xx status. */
export class SavedFoodApiError extends Error {
  readonly status: number;
  constructor(status: number, message: string) {
    super(message);
    this.name = "SavedFoodApiError";
    this.status = status;
  }
}

function authHeaders(session: SavedFoodSession): Record<string, string> {
  return {
    Authorization: `Bearer ${session.token}`,
    "Content-Type": "application/json",
    Accept: "application/json",
  };
}

function savedFoodsUrl(session: SavedFoodSession): string {
  return `${session.baseUrl}/api/users/${encodeURIComponent(session.userId)}/saved-foods`;
}

async function readError(
  response: Response,
  action: string,
): Promise<SavedFoodApiError> {
  const message =
    response.status === 401
      ? "Your session has expired. Sign in again to keep saving."
      : response.status === 404
        ? "We couldn't find that saved food."
        : response.status === 422
          ? "That food couldn't be saved. Check the values and try again."
          : `Could not ${action} (status ${response.status}).`;
  return new SavedFoodApiError(response.status, message);
}

/**
 * Save one corrected food for the caller, recording their typed phrase as an
 * alias. Returns the stored saved food carrying its nutrition snapshot.
 * Cross-user saves fail closed as 404.
 */
export async function saveFood(
  session: SavedFoodSession,
  request: SaveFoodRequest,
  fetchImpl: typeof fetch = fetch,
): Promise<SavedFoodDTO> {
  const response = await fetchImpl(savedFoodsUrl(session), {
    method: "POST",
    headers: authHeaders(session),
    body: JSON.stringify(request),
  });
  if (!response.ok) {
    throw await readError(response, "save your food");
  }
  return (await response.json()) as SavedFoodDTO;
}

/**
 * Search the caller's saved foods and aliases by normalized prefix/contains.
 * Each result carries its stored nutrition so the client can apply it directly
 * without re-estimating. Cross-user searches fail closed as 404.
 */
export async function searchSavedFoods(
  session: SavedFoodSession,
  query: string,
  fetchImpl: typeof fetch = fetch,
): Promise<SavedFoodSearchResponse> {
  const url = `${savedFoodsUrl(session)}?q=${encodeURIComponent(query)}`;
  const response = await fetchImpl(url, {
    method: "GET",
    headers: authHeaders(session),
  });
  if (!response.ok) {
    throw await readError(response, "search your saved foods");
  }
  return (await response.json()) as SavedFoodSearchResponse;
}

File: backend/app/routers/saved_foods.py
"""Saved-foods + typeahead routes (FTY-052).

``POST /api/users/{user_id}/saved-foods`` deliberately saves a corrected food and
the phrase the user typed; ``GET /api/users/{user_id}/saved-foods`` returns that
user's saved foods matching a query by normalized prefix/contains so the mobile
client (FTY-053) can re-apply stored nutrition without re-estimating.

The ``{user_id}`` path is explicit so object-level ownership is checked on every
access. A cross-user save or search fails closed: the service raises
``SavedFoodForbidden`` and this router renders ``404`` — another user's foods are
never written under, returned, nor confirmed to exist. The typed alias and query
text are never logged.
"""

from __future__ import annotations

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.orm import Session

from app.db import get_session
from app.deps import CurrentUser
from app.schemas.saved_foods import (
    MAX_QUERY_LENGTH,
    SavedFoodDTO,
    SavedFoodSearchResponse,
    SaveFoodRequest,
)
from app.services import saved_foods as saved_foods_service
from app.services.saved_foods import SavedFoodForbidden

router = APIRouter(prefix="/api/users", tags=["saved-foods"])

_NOT_FOUND = HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="saved food not found")


@router.post(
    "/{user_id}/saved-foods",
    response_model=SavedFoodDTO,
    status_code=status.HTTP_201_CREATED,
)
def save_food(
    user_id: uuid.UUID,
    payload: SaveFoodRequest,
    current_user: CurrentUser,
    session: Annotated[Session, Depends(get_session)],
) -> SavedFoodDTO:
    """Save one corrected food for the caller, mapping their typed phrase to it.

    Creates one saved food from the corrected nutrition snapshot and one alias for
    the originating phrase. Cross-user saves fail closed as ``404``.
    """

    try:
        saved_food = saved_foods_service.save_food(
            session,
            user_id,
            current_user,
            payload.name,
            payload.phrase,
            payload.nutrition,
        )
    except SavedFoodForbidden as exc:
        raise _NOT_FOUND from exc
    return SavedFoodDTO.model_validate(saved_food)


@router.get("/{user_id}/saved-foods", response_model=SavedFoodSearchResponse)
def search_saved_foods(
    user_id: uuid.UUID,
    current_user: CurrentUser,
    session: Annotated[Session, Depends(get_session)],
    q: Annotated[
        str,
        Query(
            min_length=1,
            max_length=MAX_QUERY_LENGTH,
            description="Typeahead query; matched against saved-food names and aliases "
            "by normalized prefix/contains.",
        ),
    ],
) -> SavedFoodSearchResponse:
    """Return the caller's own saved foods matching ``q`` by normalized contains.

    Each result carries its stored nutrition so the client applies it directly. A
    cross-user search fails closed as ``404``.
    """

    try:
        items, limit = saved_foods_service.search_saved_foods(session, user_id, current_user, q)
    except SavedFoodForbidden as exc:
        raise _NOT_FOUND from exc
    return SavedFoodSearchResponse(
        items=[SavedFoodDTO.model_validate(item) for item in items],
        limit=limit,
    )
